### PR TITLE
Fix/export 3d scene

### LINKED ISSUE #57437: Exporting a 3D scene results in a broken obj file
### What is the bug or the crash?

The native export 3d scene feature of QGIS results in a broken 3d model (distorted terrain and extruded buildings) when displaying it in Blender or MeshLab. 

For instance, the following scene looks like this in the 3d view : 
image

When displaying the exported version of it in Blender it looks like this: 
image

The terrain has been distorted and some of the buildings have missing pieces. 

### Steps to reproduce the issue

1. Choosing a raster DEM as terrain 
2. Choosing a vector layer to use as extruded buildings 
3. Exporting the scene using the export option in the 3d view : 
image
4. Displaying the exported 3d scene in a 3d visualization software 

### Versions

Version de QGIS
3.36.1-Maidenhead
Révision du code
3e589453
Version de Qt
5.15.3
Version de Python
3.9.18
Version de GDAL/OGR
3.8.4
Version de Proj
9.3.1
Version de la base de données du registre EPSG
v10.098 (2023-11-24)
Version de GEOS
3.12.1-CAPI-1.18.1
Version de SQLite
3.41.1
Version de PDAL
2.6.0
Version du client PostgreSQL
16.2
Version de SpatiaLite
5.1.0
Version de QWT
6.1.6
Version de QScintilla2
2.13.4
Version de l'OS
Windows 10 Version 2009

Extensions Python actives
BoundingBox
2.1
citygen
0.3
firstaid
3.1.0
idg
0.2.4
landxmlconvertor
0.2.0
model_generator_opendata
0.1
pluginbuilder3
3.2.1
plugin_reloader
0.9.4
qgis-maptiler-plugin
3.3
Qgis2threejs
2.7.3
quick_map_services
0.19.34
STL_Generator
1.0.0
valuetool
3.0.19
db_manager
0.1.20
grassprovider
2.12.99
MetaSearch
0.3.6
processing
2.12.99

### Supported QGIS version

- [x] I'm running a supported QGIS version according to [the roadmap](https://www.qgis.org/en/site/getinvolved/development/roadmap.html#release-schedule).

### New profile

- [x] I tried with a new [QGIS profile](https://docs.qgis.org/latest/en/docs/user_manual/introduction/qgis_configuration.html#working-with-user-profiles)

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by benoitdm-oslandia**
This PR fixes issues with exporting 3D scene (see #57437) by improving the way OBJ objects are saved (sparse faces/vertexes) and by improving the scene exporter (fix exporting unique feature).

fixes #57437 

Disclaimer: this PR introduces new data files to check against and may increase the git repository size by 5M

### CODE REVIEW

File: src/3d/qgs3dexportobject.cpp [hunk 1]
```diff
@@ -117,7 +117,9 @@ void Qgs3DExportObject::saveTo( QTextStream &out, float scale, const QVector3D &
     out << "s off\n";
 
   // Construct vertices
-  for ( const unsigned int vertice : mIndexes )
+  // As we can have holes in the face list and we only write vertices from theses faces
```
Comment (nyalldawson) [inline src/3d/qgs3dexportobject.cpp:120]:
```suggestion
  // As we can have holes in the face list and we only write vertices from these faces
```

File: src/3d/qgs3dsceneexporter.cpp [hunk 2]
```diff
@@ -463,25 +485,41 @@ QVector<Qgs3DExportObject *> Qgs3DSceneExporter::processInstancedPointGeometry(
   for ( Qt3DQGeometry *geometry : geometriesList )
   {
     Qt3DQAttribute *positionAttribute = findAttribute( geometry, Qt3DQAttribute::defaultPositionAttributeName(), Qt3DQAttribute::VertexAttribute );
-    Qt3DQAttribute *indexAttribute = nullptr;
-    for ( Qt3DQAttribute *attribute : geometry->attributes() )
+    Qt3DQAttribute *indexAttribute = findAttribute( geometry, "", Qt3DQAttribute::IndexAttribute );
```
Comment (nyalldawson) [inline src/3d/qgs3dsceneexporter.cpp:488]:
```suggestion
    Qt3DQAttribute *indexAttribute = findAttribute( geometry, QString(), Qt3DQAttribute::IndexAttribute );
```

File: src/3d/qgs3dsceneexporter.cpp [hunk 3]
```diff
@@ -683,23 +749,20 @@ QVector<Qgs3DExportObject *> Qgs3DSceneExporter::processLines( Qt3DCore::QEntity
...
-    Qt3DQAttribute *indexAttribute = nullptr;
-    for ( Qt3DQAttribute *attribute : geom->attributes() )
-    {
-      if ( attribute->attributeType() == Qt3DQAttribute::IndexAttribute )
-      {
-        indexAttribute = attribute;
-        break;
-      }
-    }
+    Qt3DQAttribute *indexAttribute = findAttribute( geom, "", Qt3DQAttribute::IndexAttribute );
```
Comment (nyalldawson) [inline src/3d/qgs3dsceneexporter.cpp:752]:
```suggestion
    Qt3DQAttribute *indexAttribute = findAttribute( geom, QString(), Qt3DQAttribute::IndexAttribute );
```

File: src/3d/qgs3dsceneexporter.cpp [hunk 4]
```diff
@@ -351,12 +353,32 @@ void Qgs3DSceneExporter::parseFlatTile( QgsTerrainTileEntity *tileEntity, const
 
   // Generate vertice data
   Qt3DQAttribute *positionAttribute = tileGeometry->positionAttribute();
+  if ( positionAttribute == nullptr )
+  {
+    QgsDebugError( QString( "Will not export '%1' has geometry with no position attribute!" ).arg( layerName ) );
```
Comment (nyalldawson) [inline src/3d/qgs3dsceneexporter.cpp:358]:
```suggestion
    QgsDebugError( QString( "Cannot export '%1' - geometry has no position attribute!" ).arg( layerName ) );
```

Comment (nyalldawson) [reply src/3d/qgs3dsceneexporter.cpp:358]: Elsewhere too

**Comment by nyalldawson** [comment]
How are these test results going to be useful for anyone to diagnose? Looking at the failure here it's just one random hex string vs another. :man_shrugging:

File: src/3d/qgs3dsceneexporter.cpp (hunk 4 above)
Comment (benoitdm-oslandia) [reply src/3d/qgs3dsceneexporter.cpp:358]: ok

File: src/3d/qgs3dsceneexporter.cpp (hunk 3 above)
Comment (benoitdm-oslandia) [reply src/3d/qgs3dsceneexporter.cpp:752]: ok

File: src/3d/qgs3dsceneexporter.cpp (hunk 2 above)
Comment (benoitdm-oslandia) [reply src/3d/qgs3dsceneexporter.cpp:488]: ok

File: src/3d/qgs3dexportobject.cpp (hunk 1 above)
Comment (benoitdm-oslandia) [reply src/3d/qgs3dexportobject.cpp:120]: ok

**Comment by benoitdm-oslandia** [comment]
> How are these test results going to be useful for anyone to diagnose? Looking at the failure here it's just one random hex string vs another. 🤷‍♂️

under the hex string you have got the blocks in text format. Here, we have a difference in the generated normal at the 6th digit after 0.

May be I should add an option the define the precision of saved data.

**Comment by benoitdm-oslandia** [comment]
@nyalldawson all green! Is that look good to you now?

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
@benoitdm-oslandia I'll squash this so we don't end up with the outdated versions of the large test files

**Comment by benoitdm-oslandia** [comment]
> @benoitdm-oslandia I'll squash this so we don't end up with the outdated versions of the large test files

arrg, so bad, we definitively need an `autosquash` process to handle properly the `fixup` git commits!! 

I did not force-push my previous changes with multiple `fixup` commits in order to reduce the difficulty of the review process. At merge time the `fixup` commits can be [easily squashed](https://fle.github.io/git-tip-keep-your-branch-clean-with-fixup-and-autosquash.html) with a `git rebase -i upstream/master --autosquash`

May be we can use one action like this [one](https://github.com/marketplace/actions/block-autosquash-commits)?

**Comment by benoitdm-oslandia** [comment]
Thanks for the review and the merge by the way!!!

**Comment by nyalldawson** [comment]
@benoitdm-oslandia 

It's easy enough to do from github -- it's just a different button when merging. I don't think there's any need for a workflow or any added complexity here :)

**Pull request merged**